torch_xla/core/xla_model.py

```diff
@@ -846,6 +940,101 @@ def reduce_scatter(reduce_type,
                     f"given {type(input)}.")
 
 
+def reduce_scatter_bucketized(reduce_type,
+                              input_list,
+                              scale,
+                              scatter_dim,
+                              shard_count,
+                              groups=None,
+                              output=None,
+                              pin_layout=True,
+                              bucket_cap_mb=160):
+  """Performs a XLA `ReduceScatter()` operation on a list of tensors (bucketized).
+
+  See: https://www.tensorflow.org/xla/operation_semantics#reducescatter
+
+  Args:
+    see reduce_scatter for reduce_type, scale, scatter_dim, shard_count, groups, pin_layout
+    input_list: List of input tensors
+    output: Optional list of output torch.Tensor
+    bucket_cap_mb: Number of MegaBytes of the tensor bucket to fill before doing all-gather.
+
+  Returns:
+    A list of `torch.Tensors` with all the values reduced across replicas. Each process
+    gets a shard split along the `scatter_dim`. All other dimensions are
+    the same as the input.
+  """
+  token, devctx = _get_all_reduce_token()
+
+  if not isinstance(input_list, list) or any(
+      not isinstance(v, torch.Tensor) for v in input_list):
+    raise TypeError(
+        f"`input_list` needs to be a list of Tensors, but given {type(input_list)}."
+    )
+  if output != None:
+    if not isinstance(output, list) or any(
+        not isinstance(v, torch.Tensor) for v in output):
+      raise TypeError(
+          f"`output` needs to be a list of Tensors, but given {type(output)}.")
+    if len(output) != len(input_list):
+      raise ValueError("`output` length doesn't match `input_list` length: "
+                       f"{len(output)} vs {len(input_list)}.")
+
+  def _reduce_scatter_coalesced(_input_list, _output_list=None):
+    return reduce_scatter(
+        reduce_type=reduce_type,
+        input=_input_list,
+        scale=scale,
+        scatter_dim=scatter_dim,
+        shard_count=shard_count,
+        groups=groups,
+        output=_output_list,
+        pin_layout=pin_layout)
```

Comment:
is this the only difference between `reduce_scatter_bucketized ` and `all_gather_bucketized`, if it is we can even further refactor the implementation of these two methods.

Reply:
refactored